lib/app.js

```diff
@@ -41,8 +43,29 @@ if (otpConfig.customIcons) {
   })
 }
 
-if (jsConfig.applyCustomIconLogic) {
-  jsConfig.applyCustomIconLogic(customIcons)
+const BaseModeIcon = jsConfig.BaseModeIcon || ClassicModeIcon
+
+// Build a ModeIcon component from a base ModeIcon component
+// and from the customIcons set (they can't be imported from config.js).
+// If there is an exact-matching icon in customIcons, use it.
+const ModeIcon = ({ mode, ...props }) => {
+  if (!mode) return null;
+
+  const CustomIcon = customIcons[mode]
+  if (CustomIcon) return React.cloneElement(CustomIcon, {...props})
+
+  return <BaseModeIcon mode={mode} {...props} />
+}
+
+// Build the LegIcon component, including custom icons for some legs if any.
+const LegIcon = ({ leg, ...props }) => {
+  const legIconName = jsConfig.getIconNameForLeg ? jsConfig.getIconNameForLeg(leg) : null
+
+  let CustomIcon
+  if (legIconName) CustomIcon = customIcons[legIconName]
+  if (CustomIcon) return React.cloneElement(CustomIcon, {...props})
+
+  return <BaseLegIcon leg={leg} ModeIcon={ModeIcon} {...props} />
 }
```

Comment:
I'm not too fond of divorcing the ModeIcon and LegIcon like this. Why not just define them both in `config.js` and remove `getIconNameForLeg`?